### PR TITLE
382 parser warnings

### PR DESCRIPTION
**Pull request by hammerheadlemon**
WIP PR reference #382.

Uses a decorator to allow new "dashed" semantics (e.g. "--color-owner").

Does not decorate work with the core parser as yet, but does so with color/reminder parsers.

Very grateful for comments and feedback!

### CODE REVIEW

File: gcalcli/gcalcli.py [hunk 1]
```diff
@@ -1632,7 +1648,7 @@ def main():
         else:
             tmp_argv = argv
 
-        (parsed_args, junk) = parser.parse_known_args(tmp_argv)
+        (parsed_args, unparsed) = parser.parse_known_args(tmp_argv)
```
Comment (lasers) [inline gcalcli/gcalcli.py]: This could be in a separate PR because it's a bugfix. Users on `-git` will have to wait days for this fix.

Comment (hammerheadlemon) [reply gcalcli/gcalcli.py]: That was already fixed in my PR I think.

Comment (hammerheadlemon) [reply gcalcli/gcalcli.py]: @laser - Ah, I see what you're saying. This is a current bug in master, you're right. @jcrowgey - I'll do another PR isolating this fix

Comment (hammerheadlemon) [reply gcalcli/gcalcli.py]: Fixed in PR #413.

**Comment by hammerheadlemon** [comment]
Am pushing a fix to this PR - currently, we've lost the "--help" text as I didn't implement that in `COLOR_PARSER_OPTIONS`. Will push a fix later.

File: gcalcli/gcalcli.py [hunk 2]
```diff
@@ -1619,10 +1621,27 @@ def run_add_prompt(parsed_args, printer):
             parsed_args.reminders.append(str(n) + ' ' + m)
 
 
+def _warn_deprecated_opts(argv):
```
Comment (jcrowgey) [inline gcalcli/gcalcli.py]: The issue with this implementation is that it's not robust to = syntax: `--color_owner=yellow` won't be in `ALL_DEPRECATED_OPTS`.

**Comment by jcrowgey** [comment]
Take a look at this branch: https://github.com/insanum/gcalcli/tree/382_parser_warnings

I pushed one commit on top of what you've got here.  I allow the decorator to take a param, so that the decorated functions can say which data they want to access and we don't have to have a big if/else tree in the decorator.

I think this is probably on the right track.  I'll pick this up again in a day or two if you don't beat me to it :)

Thanks again for the help.

File: gcalcli/gcalcli.py (hunk 2 above)
Comment (jcrowgey) [reply gcalcli/gcalcli.py]: We could do something with `.startswtih`, I guess.  That's not very elegant but this code will be excised in a future version, so I may be able to cope with that.

**Comment by hammerheadlemon** [comment]
No time this morning but your commit looks good, and note your comments re the "=" syntax. Will hopefully get a chance to look again at this over the day or two, including the the test.  Thanks again for your feedback.

**Comment by jcrowgey** [comment]
I've updated the 382_parser_warnings branch in this repo with an approach that removes the _warn function from gcalcli.py and which uses custom Argparse actions to implement the deprecation parsing.

I'd love any thoughts and feedback on it.  I still want to deprecate the options to the subcommands which don't have their own get_parser function, and are therefore not readily treatable with the decorator approach.

**Comment by hammerheadlemon** [comment]
Unfortunately, I'm seriously pushed for time this week so unable to dig into this in detail. However I've updated by branch and taken a look. Like the idea of subclassing the bits of `argparse` to make this work without relying on the decorator. I often find that using decorators is a bit of a paradox - one one hand it makes it clear what the intention is, at the cost of making the code slightly more complex. In an ideal world, this wouldn't be necessary but I think it's a good contingency to solve this particular problem. Perhaps flag future work on tidying up the CLI options for after 4.0?

**Comment by lasers** [comment]
I wanted to try this. https://github.com/lasers/gcalcli/tree/standardize-cli-options

**Comment by jcrowgey** [comment]
Ok, thanks to everyone in this thread for their help.  @lasers I really liked your loop in the `get_color_parser` function so I merged your commit and then edited it a bit.

One final note on this topic, all gcalcli options are now in alignment with standard cli syntax.  The existing exceptions from form oauth2client.tools which provides a parser that we import.  Those will have to stay for now.
